# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
#Lambda x: x + 10
#regular named function
def add_ten(x):
    return x + 10
print(add_ten(4))

#Lambda with variable
add_something = lambda x: x+9

print(add_something(7))
#calling Lambda without variable
print((lambda y: y * 4)(8))


14
16
32


#### Saving to a Variable

In [11]:
x = lambda num: num * 10
#def x(num):
# return num * 10
print(x(10))

y = lambda name: print(name)
y('Boba Fett')

z = lambda num: int(num) + 80
print(z('40'))

print(z)
#print(add_num)

100
Boba Fett
120
<function <lambda> at 0x00000224DA1E04C0>


#### Multiple Inputs

In [3]:
#multiple inputs with no variable
print((lambda x, y, z: x*y*z)(3,5,8))

#multiple inputs with a variable name
x_test = lambda x,y,z: x * y + z

print(x_test(5, 8, 23))

string_test = lambda verb, adjective, noun: f"It is a {adjective} {noun} and I {verb} to the store."
#
print(string_test)

120
63
<function <lambda> at 0x000002CE6143F670>


#### Passing a Lambda into a Function

In [3]:
def multiply (f, num):
    """
    f expects a lambda function
    num expects a number
    
    """
    return f(num)
print(multiply(lambda x: x*4, 6))

def multiply2(num):
    f = lambda x: x * 4
    return f(num)
print(multiply2(7))

24
28


#### Returning a Lambda from a Function

In [4]:
# regular defined function
def multiply_test(num):
    return num * 4

# function within a function
def returnFunc():
    test = 4
    def multiply(num):
        return num * test
    return multiply
func_return = returnFunc()

#print(func_return)

print(func_return(8))

#Lambda function returned from a regular function
def return_lamb(b, c):
    return lambda x, a: x + a + b + c

r_lamb = return_lamb(4,5)
print(return_lamb(4,5)(7,8))

print(r_lamb(9, 10))

32
24
28


#### If Statements within Lambdas

In [23]:
# Lambda x: True if (condition) else False
#Electric Company
#Create a function that given a list which represents street lights given as a parameter(l_street), determine if an outage has occurred. A street with a total number of "F" greater than or equal to 2 returns "Outage", anything below returns "Power"
#Example Input: [ 'T', 'F', 'F', 'F' ]
#Example Output: "Outage"




is_there_power = lambda arr: "Outrage" if arr.count('F') >= 2 else "Power"

print(is_there_power(['T', 'F', 'F', 'F']))

true_power = lambda arr: False if arr.count("F") >= 2 else true
print(true_power(['T', 'F', 'F', 'F']))

arr = ['T', 'F', 'F', 'F']
true_list = [True if x == "T" else False for x in arr]
print(true_list)


f_condition = lambda num: num * 2 if num > 10 else num + 2

print(f_condition(8))

print(f_condition(10))

print(f_condition(12))


Outrage
False
[True, False, False, False]
10
12
24


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in if they are less than 15, otherwise square them. Assign the anonymous function to a variable 'f'.</p>

In [27]:
func = lambda x: x**3 if x < 15 else x**2
print(func(10))   
print(func(20))   


1000
400


In [28]:
f = lambda x: x**3 if x < 15 else x**2
print(f(10))
print(f(20))

1000
400


In [29]:
f = lambda num: num ** 3 if num < 15 else num ** 2

print(f(20))


400


In [30]:
f = lambda x: x**2
print(f(5))

25


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [33]:
# map(function(iterable<list, dict, tuple, set))
# can use with regular named functions but lambdas work nicely as well
def square_nums(num):
    return num ** 2

arr = [2,30,4,15,6,17]

squared_list = []
for x in arr:
    squared_list.append(square_nums(x))
print(squared_list)

def squared(num):
    if num < 10:
        return num ** 2
    else:
        return num
squared_nums_map = list(map(squared, arr))

print(squared_nums_map)




[4, 900, 16, 225, 36, 289]
[4, 30, 16, 15, 36, 17]


In [38]:
arr = [2,30,4,15,6,17]

arr2 = [7,10,4,16,87,91]

def squared2(num, num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 ** 2
    else:
        return num, num2
squared_nums_map = list (map(squared2, arr, arr2))
print(squared_nums_map)

[(4, 49), (30, 10), (16, 16), (15, 16), (6, 87), (17, 91)]


#### Using Lambda's with Map

In [40]:
arr = [2,30,4,15,6,17]

arr2 = [7,10,4,16,87,91]
#map(Lambda x: x+2, iterable)
#using Lambda in map happens in one line... usually

squared_nums_lamb = list(map(lambda x, y: (x ** 2, y**2) if x < 10 and y < 10 else (x,y), arr, arr2))
print(squared_nums_lamb)


[(4, 49), (30, 10), (16, 16), (15, 16), (6, 87), (17, 91)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [ ]:
arr = [2,30,4,15,6,17]

f = lambda squared_list = []
for x in arr:
    squared_list.append(square_nums(x))
print(squared_list)

def squared(num):
    if num < 10:
        return num ** 2
    else:
        return num
squared_nums_map = list(map(squared, arr))

print(squared_nums_map)
#not done



In [41]:
arr = [2, 30, 4, 15, 6, 17]
double = list(map(lambda x: x * 2 - 1, arr))
print(double)

[3, 59, 7, 29, 11, 33]


In [42]:
new_map = list(map(lambda x: x*2-1, arr))
print(new_map)

[3, 59, 7, 29, 11, 33]


In [43]:
# y = lambda x: x*2 - 1
math_double_substractOne = list(map(lambda x: x*2 - 1, arr))
print(math_double_substractOne)

[3, 59, 7, 29, 11, 33]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [45]:
names =['Bob', 'Jason', 'Vince', 'Evan', 'Angela']
def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
    
new_names_list = list(filter(a_names, names))
print(new_names_list)


numbers = [1,2,3,4,5,6,7,8,9,10]

def even_nums(number):
    if number % 2 == 0:
        return True
    else:
        return False
    
even_numbers = list(filter(even_nums, numbers))
print(even_numbers)

['Angela']
[2, 4, 6, 8, 10]


#### Using Lambda's with Filter()

In [49]:
names =['Bob', 'Jason', 'Vince', 'Evan', 'Angela']
a_names_list = list(filter(lambda name: True if name[0].lower == 'a' else False, names))
print(a_names_list)


numbers = [1,2,3,4,5,6,7,8,9,10]


even_numbers = list(filter(lambda num: True if num % 2 == 0 else False, numbers))

print(even_numbers)

[]
[2, 4, 6, 8, 10]


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [1]:
from statistics import mean

nums = [2, 7, 4.2, 1.6, 9, 4.4, 5.9]

print(mean(nums))


4.871428571428572


In [2]:
below_mean = list(filter(lambda a: True if a <= mean(nums) else False, nums))
print(below_mean)

[2, 4.2, 1.6, 4.4]


In [3]:
from statistics import mean

nums = [2, 7, 4.2, 1.6, 9, 4.4, 5.9]

print(mean(nums))
print(list(filter(lambda num: num if num < mean(nums) else False, nums)))

4.871428571428572
[2, 4.2, 1.6, 4.4]


In [4]:
print(list(filter(
(lambda x: True if x <= mean(nums) else False),nums
)))

[2, 4.2, 1.6, 4.4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [12]:
from functools import reduce
#reduce(function, iterable)
list_1 = [2,4,6,8,10]

def add_nums(num1,num2):
    return num1 + num2


result_add = reduce(add_nums, list_1)
print(result_add)

#subtract a list of numbers
def subtract_nums(num1, num2):
    return num1 - num2
result_sub = reduce(subtract_nums, list_1)
print(result_sub)

30
-26


#### Using Lambda's with Reduce()

In [ ]:
result_lamb = reduce(lambda x, y: x + y, list_1)
print(result_lamb)

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [ ]:
list_1 = [2,4,6,8,10]

In [13]:
from functools import reduce

numbers = [2, 3, 4, 5, 6]

result = reduce(lambda x, y: x * y, numbers)

print(result) # Output: 720


720


In [15]:
list_1 = [2,4,6,8,10]
multiply_lamb = reduce (lambda x, y: x + y, list_1, 10)
print(multiply_lamb)

40


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

#### Writing a Factorial Function

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

## Generators #Pop Interview question<br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [1]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start #yield keyword there be a generatorrrrrr
        start += step
        
for i in my_range(20, start = 2):
    my_generator_value = i
    print(my_generator_value)
    



2
4
6
8
10
12
14
16
18


#### next() keyword
<p>Returns next element from the list, if not present prints the default value. If default value is not present, raises the StopIteration error.</p>

In [ ]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start #yield keyword there be a generatorrrrrr
        start += step
        

#### Creating a Generator Object
<p>The generator object will store an iterable on which we can call the next() method.</p>

In [5]:
generator_object = my_range(20, start = 2)

In [6]:
#next keyword will iterate through our generator object and yield a new value eahc time
next(generator_object)

2

<p>Use try and except to print graceful message instead of the StopIteration error:

In [7]:
try:
    print(next(generator_object))
    if next(generator_object) == 12:
        print("supa dupa")
    else:
        print("it okey")
except:
    print("You're out of range")

4
it okey


#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [ ]:
numarg = **
    print()

In [8]:
#chatgpt example

def squared_generator(n):
    while n > 0:
        yield n ** 2
        n -= 1

for num in squared_generator(5):
    print(num)


25
16
9
4
1


In [9]:
def squared(stop, start, step = 1):
    while start >= stop:
        yield start ** 2
        start -= step
gen = squared(0, start = 25)

try: print(next(gen))
except: print("number has reached 0")

625


In [10]:
def my_range(stop, start, step):
    while start >= 0:
        yield start**2
        start -= step
for i in my_range(0,10,1):
    my_generator_value = i
    print(my_generator_value)

100
81
64
49
36
25
16
9
4
1
0


In [ ]:
try:
    print()

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [5]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
#---------------------------------------------------------------------------------
# != takes away and removes the empty strings
places = ['', 'Argentina', '', 'San Diego', 'Boston', '', 'New York', '']
result = list(filter(lambda x: x != '', places))
print(result)


['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [14]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
#-------------------------------------------------------------------------------------------------------------
# split() function splits up the names while the sort function categorizes the last names in alphabetical order starting with lowercase "a", ie aNisimov

author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield", "David hassELHOFF", "Gary A.J. Bernstein"]
author.sort(key=lambda x: x.split()[-1].lower())

print(author)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [15]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
#------------------------------------------------------------------------------------
# map function creates the conversion of celsius to farhenheit while the lambda follows the equation in the parentheses
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

converted_temps = list(map(lambda x: (x[0], (9/5)*x[1] + 32), places))

print(converted_temps) 


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 *Interview Question Example*<br>
<p>Create a generator function that individually returns each movie genre back from the list</p>




In [24]:
genres = [("adventure"), ("drama"), ("horror"), ("comedy"), ("action"), ("romance")]
def genre_generator(genres):
    for genre in genres:
        yield genre

my_genres = genre_generator(genres)

print(next(my_genres))
print(next(my_genres))
print(next(my_genres))
print(next(my_genres))
print(next(my_genres))
print(next(my_genres))


adventure
drama
horror
comedy
action
romance
